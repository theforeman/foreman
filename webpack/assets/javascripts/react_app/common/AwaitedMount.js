import PropTypes from 'prop-types';
import { useState, useEffect, useCallback } from 'react';
import store from '../redux';
import componentRegistry from '../components/componentRegistry';

// Mounts a component after all plugins have been imported to make sure that all plugins are available to the component
export const AwaitedMount = ({ component, data, flattenData }) => {
  const [mounted, setMounted] = useState(false);
  const [mountedComponent, setMountedComponent] = useState(null);
  const mountComponent = useCallback(async () => {
    if (componentRegistry.registry[component]) {
      setMounted(true);
      setMountedComponent(
        componentRegistry.markup(component, {
          data,
          store,
          flattenData,
        })
      );
    } else {
      // eslint-disable-next-line no-console
      console.debug(
        `Component not found: ${component}. The script for the component might not have been loaded yet.`
      );
    }
  }, [component, data, flattenData]);
  useEffect(() => {
    document.addEventListener('loadPlugin', mountComponent);
    return () => {
      window.removeEventListener('loadPlugin', mountComponent);
    };
  }, [mountComponent]);
  useEffect(() => {
    mountComponent();
  }, [mountComponent]);
  return mounted ? mountedComponent : null;
};

AwaitedMount.propTypes = {
  component: PropTypes.string.isRequired,
  data: PropTypes.object,
  flattenData: PropTypes.bool,
};
AwaitedMount.defaultProps = {
  data: {},
  flattenData: false,
};
