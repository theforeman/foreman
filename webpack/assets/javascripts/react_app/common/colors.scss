// colors
$pf-black-200: #ededed;
$pf-red-200: #a30000;
$pf-red-100: #c00;
$pf-orange-400: #ec7a08;
$pf-gold-400: #f0ab00;
$pf-green-400: #3f9c35;
$pf-green-600: #1e4f18;
$pf-blue-400: #0088ce;
$pf-color-white: #fff;
$pf-border-gray: #d1d1d1;

$switcher-max-height: 400px;
$switcher-max-width: 200px;
$switcher-min-width: 200px;

$nav-pf-vertical-active-icon-color: white;
$nav-pf-vertical-active-bg-color: #026991;
$nav-pf-vertical-secondary-active-bg-color: #024d6c;
$nav-pf-vertical-icon-color: #c7c7c7;
$nav-pf-vertical-secondary-indicator-color: #d1d1d1;
$nav-pf-vertical-bg-color: #024d6c;
