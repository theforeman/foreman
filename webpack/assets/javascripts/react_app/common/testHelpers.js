import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

export default {
  mockStorage: () => {
    const storage = {};

    return {
      setItem: (key, value) => {
        storage[key] = value || '';
      },
      getItem: key => (key in storage ? storage[key] : null),
      removeItem: key => {
        delete storage[key];
      },
      get length() {
        return Object.keys(storage).length;
      },
      key: i => {
        const keys = Object.keys(storage);

        return keys[i] || null;
      },
    };
  },
};

// a helper method for invoking a class method (for unit tests)
// obj = a class
// func = a tested function
// objThis = an object's this
// arg = function args

export const classFunctionUnitTest = (obj, func, objThis, args) =>
  obj.prototype[func].apply(objThis, args);

/**
 * Shallow render a component multipile times with fixtures
 * @param  {ReactComponent} Component Component to shallow-render
 * @param  {Object}         fixtures  key=fixture description, value=props to apply
 * @return {Object}                   key=fixture description, value=shallow-rendered component
 */
export const shallowRenderComponentWithFixtures = (Component, fixtures) =>
  Object.entries(fixtures).map(([description, props]) => ({
    description,
    component: shallow(<Component {...props} />),
  }));

/**
 * Test a component with fixtures and snapshots
 * @param  {ReactComponent} Component Component to test
 * @param  {Object}         fixtures  key=fixture description, value=props to apply
 */
export const testComponentSnapshotsWithFixtures = (Component, fixtures) =>
  shallowRenderComponentWithFixtures(Component, fixtures).forEach(
    ({ description, component }) =>
      it(description, () => expect(toJson(component)).toMatchSnapshot())
  );

/**
 * run an action (sync or async) and except the results to much snapshot
 * @param  {Function}  runAction  Action runner function
 * @return {Promise}
 */
export const testActionSnapshot = async runAction => {
  const actionResults = runAction();

  // if it's an async action
  if (typeof actionResults === 'function') {
    const dispatch = jest.fn();
    await actionResults(dispatch);

    expect(dispatch.mock.calls).toMatchSnapshot();
  } else {
    expect(actionResults).toMatchSnapshot();
  }
};

/**
 * Test actions with fixtures and snapshots
 * @param  {Object} fixtures key=fixture description, value=action runner function
 */
export const testActionSnapshotWithFixtures = fixtures =>
  Object.entries(fixtures).forEach(([description, runAction]) =>
    it(description, () => testActionSnapshot(runAction))
  );

/**
 * Test a reducer with fixtures and snapshots
 * @param  {Function} reducer  reducer to test
 * @param  {Object}   fixtures key=fixture description, value=props to apply
 */
export const testReducerSnapshotWithFixtures = (reducer, fixtures) => {
  const reduce = ({ state, action = {} } = {}) => reducer(state, action);
  Object.entries(fixtures).forEach(([description, action]) =>
    it(description, () => expect(reduce(action)).toMatchSnapshot())
  );
};

/**
 * Test selectors with fixtures and snapshots
 * @param  {Object} fixtures  key=fixture description,
 *                            value=selector runner function
 */
export const testSelectorsSnapshotWithFixtures = fixtures =>
  Object.entries(fixtures).forEach(([description, selectorRunner]) =>
    it(description, () => expect(selectorRunner()).toMatchSnapshot())
  );
