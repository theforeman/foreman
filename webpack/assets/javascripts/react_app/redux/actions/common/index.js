import API from '../../../API';

export const ajaxRequestAction = ({
  dispatch,
  requestAction,
  successAction,
  failedAction,
  url,
  item,
}) => {
  dispatch({ type: requestAction, payload: item });
  return API.get(url)
    .then(({ data }) =>
      dispatch({ type: successAction, payload: { ...item, ...data } })
    )
    .catch(error => dispatch({ type: failedAction, payload: { error, item } }));
};
