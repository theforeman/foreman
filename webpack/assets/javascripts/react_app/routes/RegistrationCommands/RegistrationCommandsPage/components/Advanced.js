import React from 'react';
import PropTypes from 'prop-types';

import ConfigParams from './fields/ConfigParams';
import Packages from './fields/Packages';
import Repository from './fields/Repository';
import TokenLifeTime from './fields/TokenLifeTime';
import UpdatePackages from './fields/UpdatePackages';

const Advanced = ({
  configParams,
  setupRemoteExecution,
  setupInsights,
  handleInsights,
  handleRemoteExecution,
  jwtExpiration,
  handleJwtExpiration,
  handleInvalidField,
  packages,
  handlePackages,
  repo,
  handleRepo,
  repoGpgKeyUrl,
  handleRepoGpgKeyUrl,
  updatePackages,
  handleUpdatePackages,
  isLoading,
}) => (
  <>
    <ConfigParams
      configParams={configParams}
      setupRemoteExecution={setupRemoteExecution}
      setupInsights={setupInsights}
      handleInsights={handleInsights}
      handleRemoteExecution={handleRemoteExecution}
      isLoading={isLoading}
    />
    <Packages
      packages={packages}
      handlePackages={handlePackages}
      configParams={configParams}
      isLoading={isLoading}
    />
    <UpdatePackages
      updatePackages={updatePackages}
      handleUpdatePackages={handleUpdatePackages}
      isLoading={isLoading}
    />
    <Repository
      repo={repo}
      handleRepo={handleRepo}
      repoGpgKeyUrl={repoGpgKeyUrl}
      handleRepoGpgKeyUrl={handleRepoGpgKeyUrl}
      isLoading={isLoading}
    />
    <TokenLifeTime
      value={jwtExpiration}
      onChange={handleJwtExpiration}
      handleInvalidField={handleInvalidField}
      isLoading={isLoading}
    />
  </>
);

Advanced.propTypes = {
  configParams: PropTypes.object,
  setupRemoteExecution: PropTypes.string,
  setupInsights: PropTypes.string,
  handleInsights: PropTypes.func.isRequired,
  handleRemoteExecution: PropTypes.func.isRequired,
  jwtExpiration: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  handleJwtExpiration: PropTypes.func.isRequired,
  handleInvalidField: PropTypes.func.isRequired,
  packages: PropTypes.string,
  repo: PropTypes.string,
  repoGpgKeyUrl: PropTypes.string,
  handlePackages: PropTypes.func.isRequired,
  handleRepo: PropTypes.func.isRequired,
  handleRepoGpgKeyUrl: PropTypes.func.isRequired,
  updatePackages: PropTypes.bool,
  handleUpdatePackages: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
};

Advanced.defaultProps = {
  configParams: {},
  setupRemoteExecution: '',
  setupInsights: '',
  jwtExpiration: 4,
  packages: '',
  repo: '',
  repoGpgKeyUrl: '',
  updatePackages: false,
};

export default Advanced;
