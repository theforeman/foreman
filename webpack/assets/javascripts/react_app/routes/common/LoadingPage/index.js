import React from 'react';
import { Spinner } from 'patternfly-react';
import './loadingpage.scss';

const LoadingPage = () => (
  <div id="loading-page">
    <Spinner loading size="lg" />
  </div>
);

export default LoadingPage;
