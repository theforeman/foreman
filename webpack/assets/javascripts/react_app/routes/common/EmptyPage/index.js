import React from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from '../../../common/I18n';
import DefaultEmptyState from '../../../components/common/EmptyState';
import './emptypage.scss';

const EmptyPage = ({ message: { type, text } }) => (
  <DefaultEmptyState
    icon={type === 'error' ? 'error-circle-o' : 'add-circle-o'}
    header={type === 'error' ? __('Error') : __('No Results')}
    description={text}
  />
);

EmptyPage.propTypes = {
  message: PropTypes.shape({
    type: PropTypes.oneOf(['empty', 'error']),
    text: PropTypes.string,
  }),
};

EmptyPage.defaultProps = {
  message: PropTypes.shape({
    type: 'empty',
    text: 'No Results',
  }),
};

export default EmptyPage;
