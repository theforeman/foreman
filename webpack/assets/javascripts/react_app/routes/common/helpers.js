import { foremanUrl, urlBuilder } from '../../common/urlHelpers';

export const getControllerSearchProps = (
  controller,
  id = 'searchBar',
  canCreate = true
) => ({
  controller,
  autocomplete: {
    id,
    searchQuery: '',
    url: urlBuilder(controller, 'auto_complete_search'),
    useKeyShortcuts: true,
  },
  bookmarks: {
    url: foremanUrl('/api/bookmarks'),
    canCreate,
    documentationUrl: `4.1.5Searching`,
  },
});
