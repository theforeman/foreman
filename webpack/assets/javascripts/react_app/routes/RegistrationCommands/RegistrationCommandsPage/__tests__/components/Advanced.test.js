import { testComponentSnapshotsWithFixtures } from '../../../../../common/testHelpers';

import Advanced from '../../components/Advanced';

import { advancedComponentProps } from '../fixtures'


describe('RegistrationCommandsPage - Advanced', () => {
  testComponentSnapshotsWithFixtures(Advanced, { 'renders': advancedComponentProps });
})
