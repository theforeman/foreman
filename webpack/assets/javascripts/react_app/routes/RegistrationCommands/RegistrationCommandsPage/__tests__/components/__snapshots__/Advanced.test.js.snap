// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationCommandsPage - Advanced renders 1`] = `
<Fragment>
  <ConfigParams
    configParams={Object {}}
    handleInsights={[Function]}
    handleRemoteExecution={[Function]}
    isLoading={false}
    setupInsights=""
    setupRemoteExecution=""
  />
  <Packages
    configParams={Object {}}
    handlePackages={[Function]}
    isLoading={false}
    packages=""
  />
  <UpdatePackages
    handleUpdatePackages={[Function]}
    isLoading={false}
    updatePackages={false}
  />
  <Repository
    handleRepo={[Function]}
    handleRepoGpgKeyUrl={[Function]}
    isLoading={false}
    repo=""
    repoGpgKeyUrl=""
  />
  <TokenLifeTime
    handleInvalidField={[Function]}
    isLoading={false}
    onChange={[Function]}
    value=""
  />
</Fragment>
`;
